sql/resource_sql_endpoint_test.go

```diff
@@ -239,6 +240,45 @@ func TestResourceSQLEndpointUpdate(t *testing.T) {
 	assert.Equal(t, "d7c9d05c-7496-4c69-b089-48823edad40c", d.Get("data_source_id"))
 }
 
+// Testing the customizeDiff on clearing "health" diff is working as expected.
+func TestResourceSQLEndpointUpdateHealthNoDiff(t *testing.T) {
+	qa.ResourceFixture{
+		MockWorkspaceClientFunc: func(mwc *mocks.MockWorkspaceClient) {
+			api := mwc.GetMockWarehousesAPI()
+			api.EXPECT().Edit(mock.Anything, sql.EditWarehouseRequest{
+				Id:                 "abc",
+				Name:               "foo",
+				ClusterSize:        "Small",
+				AutoStopMins:       120,
+				MaxNumClusters:     1,
+				EnablePhoton:       true,
+				SpotInstancePolicy: "COST_OPTIMIZED",
+			}).Return(&sql.WaitGetWarehouseRunning[struct{}]{Poll: poll.Simple(getResponse)}, nil)
+			api.EXPECT().GetById(mock.Anything, "abc").Return(&getResponse, nil)
+			addDataSourceListHttpFixture(mwc)
+		},
+		Resource: ResourceSqlEndpoint(),
+		ID:       "abc",
+		InstanceState: map[string]string{
+			"name":                 "foo",
+			"cluster_size":         "Small",
+			"auto_stop_mins":       "120",
+			"enable_photon":        "true",
+			"max_num_clusters":     "1",
+			"spot_instance_policy": "COST_OPTIMIZED",
+		},
+		Update: true,
+		HCL: `
+		name = "foo"
+  		cluster_size = "Small"
+		`,
+	}.ApplyCheckDiff(t, func(diff *terraform.InstanceDiff) bool {
+		m := diff.Attributes
+		_, healthDiffExists := m["health"]
+		return !healthDiffExists
```

Comment:
This is weird. This test executes the update pathway, but we don't actually care about that. To check the diff, we don't need anything except for the config and instance state. Is this really the optimal interface for testing diff checking?

Here's what I think would be a better example of a customize diff check:
```go
qa.ResourceFixture{
  Resource: ResourceSqlEndpoint(),
  ID: "abc",
  InstanceState: ...,
  HCL: ...,
  CustomizeDiff: true,
  ExpectedDiff: map[string]*terraform.ResourceAttrDiff // Or func(*terraform.InstanceDiff) error for flexibility
}.ApplyNoError(t)
```

We don't need to test any execution when unit testing customize diff.

Reply:
A bit confused by `we don't need to test any execution` and the example you put above. I think running `ApplyNoError` still does the execution?

For the `ExpectedDiff` interface, I thought about that but I decided to put in a lambda that asserts certain properties of the diff to be true because the diff can be fairly large sometimes and specifying the whole diff manually can be less elegant than just asserting certain diffs exists

Reply:
Discussed offline, for `ExpectedDiff`, it is giving us better control over the diffs if we specify the full ResourceAttrDiff object, so going with that.